lib/instrumentation/trace-context.js

```diff
@@ -0,0 +1,109 @@
+'use strict'
+
+const crypto = require('crypto')
+
+const SIZES = {
+  version: 1,
+  traceId: 16,
+  id: 8,
+  flags: 1,
+  parentId: 8,
+
+  // Aggregate sizes
+  ids: 24, // traceId + id
+  all: 34
+}
+
+const OFFSETS = {
+  version: 0,
+  traceId: SIZES.version,
+  id: SIZES.version + SIZES.traceId,
+  flags: SIZES.version + SIZES.ids,
+
+  // Additional parentId is stored after the header content
+  parentId: SIZES.version + SIZES.ids + SIZES.flags
+}
+
+function defineLazyProp (obj, prop, fn) {
+  Object.defineProperty(obj, prop, {
+    configurable: true,
+    enumerable: true,
+    get () {
+      const value = fn()
+      Object.defineProperty(obj, prop, {
+        enumerable: true,
+        value
+      })
+      return value
+    }
+  })
+
+  return obj
+}
+
+function hexSliceFn (buffer, offset, length) {
+  return () => buffer.slice(offset, length).toString('hex')
+}
+
+function maybeHexSliceFn (buffer, offset, length) {
+  const fn = hexSliceFn(buffer, offset, length)
+  return () => {
+    const value = fn()
+    if (/[1-9a-f]/.test(value)) {
+      return value
+    }
+  }
+}
+
+// Move id into parentId region before generating new id
+function makeChild (buffer) {
+  buffer.copy(buffer, OFFSETS.parentId, OFFSETS.id, OFFSETS.flags)
+  crypto.randomFillSync(buffer, OFFSETS.id, SIZES.id)
+  return new TraceContext(buffer)
+}
+
+const bufferSymbol = Symbol('trace-context-buffer')
+
+class TraceContext {
+  constructor (buffer) {
+    this[bufferSymbol] = buffer
+    defineLazyProp(this, 'version', hexSliceFn(buffer, OFFSETS.version, OFFSETS.traceId))
+    defineLazyProp(this, 'traceId', hexSliceFn(buffer, OFFSETS.traceId, OFFSETS.id))
+    defineLazyProp(this, 'id', hexSliceFn(buffer, OFFSETS.id, OFFSETS.flags))
+    defineLazyProp(this, 'flags', hexSliceFn(buffer, OFFSETS.flags, OFFSETS.parentId))
+    defineLazyProp(this, 'parentId', maybeHexSliceFn(buffer, OFFSETS.parentId))
+  }
+
+  static create (sampled = false) {
+    const buffer = Buffer.alloc(SIZES.all)
+    crypto.randomFillSync(buffer, OFFSETS.traceId, SIZES.ids)
+    if (sampled) buffer[OFFSETS.flags] = 0x01
```

Comment:
I think the flags in the spec changed, so we should update to reflect that: https://github.com/w3c/distributed-tracing/blob/master/trace_context/HTTP_HEADER_FORMAT.md#trace-flags

Reply:
So "sampled" should mean both requested _and_ recorded?

Reply:
I'm not entirely sure. I think only `0b00000010` is sampled, but maybe @felixbarny can verify?

Reply:
The sampling probability is determined by the transaction_sample_rate configuration option. If a trace should be sampled, both the requested and the recorded flags are set. I.e. the sampler `requested` the trace to be sampled and that's why we are` record`ing it.

Reply:
@felixbarny Ok so in this case the flag should be `0b11`. And in case of an incoming request, what flag should we rely on to turn on sampling?

Reply:
For now, that would be the requested flag. If the parent propagated the requested flag, we should also set recorded to true. In the future, we might want to add rate limiting here.